src/client-side-encryption/autoEncrypter.ts

```diff
@@ -0,0 +1,641 @@
+
+    if (callback == null) {
+      throw new TypeError('Callback must be provided');
+    }
+
+    options = typeof options === 'function' ? {} : options;
+
+    // If `bypassAutoEncryption` has been specified, don't encrypt
+    if (this._bypassEncryption) {
+      callback(undefined, cmd);
+      return;
+    }
+
+    const commandBuffer = Buffer.isBuffer(cmd) ? cmd : serialize(cmd, options);
+
+    let context;
+    try {
+      context = this._mongocrypt.makeEncryptionContext(
+        MongoDBCollectionNamespace.fromString(ns).db,
+        commandBuffer
+      );
+    } catch (err) {
+      callback(err, undefined);
+      return;
+    }
+
+    // TODO: should these be accessors from the addon?
+    context.id = this._contextCounter++;
+    context.ns = ns;
+    context.document = cmd;
+
+    const stateMachine = new StateMachine({
+      ...options,
+      promoteValues: false,
+      promoteLongs: false,
+      proxyOptions: this._proxyOptions,
+      tlsOptions: this._tlsOptions
+    });
+    stateMachine.execute(this, context, callback);
+  }
+
+  /**
+   * Decrypt a command response
+   */
+  decrypt(
+    response: Uint8Array,
+    options: CommandOptions | Callback<Document>,
+    callback?: Callback<Document>
+  ) {
+    callback = typeof options === 'function' ? options : callback;
+
+    if (callback == null) {
+      throw new TypeError('Callback must be provided');
+    }
+
+    options = typeof options === 'function' ? {} : options;
+
+    const buffer = Buffer.isBuffer(response) ? response : serialize(response, options);
+
+    let context;
+    try {
+      context = this._mongocrypt.makeDecryptionContext(buffer);
+    } catch (err) {
+      callback(err, undefined);
+      return;
+    }
+
+    // TODO: should this be an accessor from the addon?
+    context.id = this._contextCounter++;
+
+    const stateMachine = new StateMachine({
+      ...options,
+      proxyOptions: this._proxyOptions,
+      tlsOptions: this._tlsOptions
+    });
+
+    // @ts-expect-error unique symbol cannot be used as an index type
+    const decorateResult = this[Symbol.for('@@mdb.decorateDecryptionResult')];
+    stateMachine.execute(this, context, function (error?: Error, result?: Document) {
+      // Only for testing/internal usage
+      if (!error && result && decorateResult) {
+        const error = decorateDecryptionResult(result, response);
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        if (error) return callback!(error);
+      }
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      callback!(error, result);
+    });
+  }
+
+  /**
+   * Ask the user for KMS credentials.
+   *
+   * This returns anything that looks like the kmsProviders original input
+   * option. It can be empty, and any provider specified here will override
+   * the original ones.
+   */
+  async askForKMSCredentials(): Promise<KMSProviders> {
+    return loadCredentials(this._kmsProviders);
```

Comment:
removed `onKmsProvidersRefresh`.  

We could consider removing this entirely and just calling `loadCredentials` directly in the state machine.

Reply:
Do we want to make this change?